Köprü: syscall olarak çağrılan fonksiyonlarda 32 bitten küçük argümanların üst baytlarını temizle.
Üreten: tools/abi-scan/scan_small_args.py --patch (elle düzenleme yapma, yeniden üret).

--- a/wine/dlls/ntdll/unix/debug.c
+++ b/wine/dlls/ntdll/unix/debug.c
@@ -388,6 +388,7 @@
  */
 NTSTATUS WINAPI NtSetDebugFilterState( ULONG component_id, ULONG level, BOOLEAN state )
 {
+    KOPRU_CLEAN_SMALL_ARG( state );
     FIXME( "component_id %#x, level %u, state %#x stub.\n", component_id, level, state );
 
     return STATUS_SUCCESS;
--- a/wine/dlls/ntdll/unix/env.c
+++ b/wine/dlls/ntdll/unix/env.c
@@ -2264,6 +2264,7 @@
  */
 NTSTATUS WINAPI NtQueryDefaultLocale( BOOLEAN user, LCID *lcid )
 {
+    KOPRU_CLEAN_SMALL_ARG( user );
     *lcid = user ? user_lcid : system_lcid;
     return STATUS_SUCCESS;
 }
@@ -2274,6 +2275,7 @@
  */
 NTSTATUS WINAPI NtSetDefaultLocale( BOOLEAN user, LCID lcid )
 {
+    KOPRU_CLEAN_SMALL_ARG( user );
     if (user) user_lcid = lcid;
     else system_lcid = lcid;
     return STATUS_SUCCESS;
@@ -2295,6 +2297,7 @@
  */
 NTSTATUS WINAPI NtSetDefaultUILanguage( LANGID lang )
 {
+    KOPRU_CLEAN_SMALL_ARG( lang );
     user_ui_language = lang;
     return STATUS_SUCCESS;
 }
--- a/wine/dlls/ntdll/unix/file.c
+++ b/wine/dlls/ntdll/unix/file.c
@@ -2858,6 +2858,8 @@
                                       FILE_INFORMATION_CLASS info_class, BOOLEAN single_entry,
                                       UNICODE_STRING *mask, BOOLEAN restart_scan )
 {
+    KOPRU_CLEAN_SMALL_ARG( single_entry );
+    KOPRU_CLEAN_SMALL_ARG( restart_scan );
     int cwd, fd, needs_close;
     enum server_fd_type type;
     struct dir_data *data;
@@ -6944,6 +6946,8 @@
                             IO_STATUS_BLOCK *io_status, LARGE_INTEGER *offset,
                             LARGE_INTEGER *count, ULONG *key, BOOLEAN dont_wait, BOOLEAN exclusive )
 {
+    KOPRU_CLEAN_SMALL_ARG( dont_wait );
+    KOPRU_CLEAN_SMALL_ARG( exclusive );
     static int warn;
     unsigned int ret;
     HANDLE handle;
@@ -7116,6 +7120,7 @@
                                              void *apc_context, IO_STATUS_BLOCK *iosb, void *buffer,
                                              ULONG buffer_size, ULONG filter, BOOLEAN subtree )
 {
+    KOPRU_CLEAN_SMALL_ARG( subtree );
     struct async_fileio_read_changes *fileio;
     unsigned int status;
     ULONG size = max( 4096, buffer_size );
@@ -7622,6 +7627,8 @@
                                BOOLEAN single_entry, void *list, ULONG list_len,
                                ULONG *index, BOOLEAN restart )
 {
+    KOPRU_CLEAN_SMALL_ARG( single_entry );
+    KOPRU_CLEAN_SMALL_ARG( restart );
     int fd, needs_close;
     NTSTATUS status;
 
--- a/wine/dlls/ntdll/unix/registry.c
+++ b/wine/dlls/ntdll/unix/registry.c
@@ -625,6 +625,8 @@
                                             IO_STATUS_BLOCK *io, ULONG filter, BOOLEAN subtree,
                                             void *buffer, ULONG length, BOOLEAN async )
 {
+    KOPRU_CLEAN_SMALL_ARG( subtree );
+    KOPRU_CLEAN_SMALL_ARG( async );
     unsigned int ret;
 
     TRACE( "(%p,%u,%p,%p,%p,%p,%p,0x%08x, 0x%08x,%p,0x%08x,0x%08x)\n",
@@ -668,6 +670,8 @@
                                    IO_STATUS_BLOCK *io, ULONG filter, BOOLEAN subtree,
                                    void *buffer, ULONG length, BOOLEAN async )
 {
+    KOPRU_CLEAN_SMALL_ARG( subtree );
+    KOPRU_CLEAN_SMALL_ARG( async );
     return NtNotifyChangeMultipleKeys( key, 0, NULL, event, apc, apc_context,
                                        io, filter, subtree, buffer, length, async );
 }
--- a/wine/dlls/ntdll/unix/security.c
+++ b/wine/dlls/ntdll/unix/security.c
@@ -186,6 +186,7 @@
  */
 NTSTATUS WINAPI NtOpenThreadToken( HANDLE thread, DWORD access, BOOLEAN self, HANDLE *handle )
 {
+    KOPRU_CLEAN_SMALL_ARG( self );
     return NtOpenThreadTokenEx( thread, access, self, 0, handle );
 }
 
@@ -196,6 +197,7 @@
 NTSTATUS WINAPI NtOpenThreadTokenEx( HANDLE thread, DWORD access, BOOLEAN self, DWORD attributes,
                                      HANDLE *handle )
 {
+    KOPRU_CLEAN_SMALL_ARG( self );
     unsigned int ret;
 
     TRACE( "(%p,0x%08x,%u,0x%08x,%p)\n", thread, access, self, attributes, handle );
@@ -223,6 +225,7 @@
 NTSTATUS WINAPI NtDuplicateToken( HANDLE token, ACCESS_MASK access, OBJECT_ATTRIBUTES *attr,
                                   BOOLEAN effective_only, TOKEN_TYPE type, HANDLE *handle )
 {
+    KOPRU_CLEAN_SMALL_ARG( effective_only );
     SECURITY_IMPERSONATION_LEVEL level = SecurityAnonymous;
     unsigned int status;
     data_size_t len;
@@ -751,6 +754,7 @@
 NTSTATUS WINAPI NtAdjustGroupsToken( HANDLE token, BOOLEAN reset, TOKEN_GROUPS *groups,
                                      ULONG length, TOKEN_GROUPS *prev, ULONG *retlen )
 {
+    KOPRU_CLEAN_SMALL_ARG( reset );
     FIXME( "%p %d %p %u %p %p\n", token, reset, groups, length, prev, retlen );
     return STATUS_NOT_IMPLEMENTED;
 }
@@ -762,6 +766,7 @@
 NTSTATUS WINAPI NtAdjustPrivilegesToken( HANDLE token, BOOLEAN disable, TOKEN_PRIVILEGES *privs,
                                          DWORD length, TOKEN_PRIVILEGES *prev, DWORD *retlen )
 {
+    KOPRU_CLEAN_SMALL_ARG( disable );
     unsigned int ret;
 
     TRACE( "(%p,0x%08x,%p,0x%08x,%p,%p)\n", token, disable, privs, length, prev, retlen );
@@ -947,6 +952,7 @@
                                             ACCESS_MASK *access_granted, NTSTATUS *access_status,
                                             BOOLEAN *onclose )
 {
+    KOPRU_CLEAN_SMALL_ARG( creation );
     FIXME( "(%s, %p, %s, %p, 0x%08x, %p, %d, %p, %p, %p), stub\n",
            debugstr_us(subsystem), handle, debugstr_us(typename), descr, access,
            mapping, creation, access_granted, access_status, onclose );
@@ -966,6 +972,7 @@
                                                   ACCESS_MASK *access_granted, NTSTATUS *access_status,
                                                   BOOLEAN *onclose )
 {
+    KOPRU_CLEAN_SMALL_ARG( creation );
     FIXME( "(%s, %p, %s, %s, %p, 0x%08x, %u, %x, %p, %u, %p, %d, %p, %p, %p), stub\n",
            debugstr_us(subsystem), handle, debugstr_us(typename), debugstr_us(objectname),
            descr, access, audit_type, flags, obj_list, list_len,
@@ -979,6 +986,7 @@
  */
 NTSTATUS WINAPI NtCloseObjectAuditAlarm( UNICODE_STRING *subsystem, HANDLE handle, BOOLEAN onclose )
 {
+    KOPRU_CLEAN_SMALL_ARG( onclose );
     FIXME( "%s %p %u: stub\n", debugstr_us(subsystem), handle, onclose );
     return STATUS_NOT_IMPLEMENTED;
 }
--- a/wine/dlls/ntdll/unix/server.c
+++ b/wine/dlls/ntdll/unix/server.c
@@ -838,6 +838,7 @@
  */
 NTSTATUS WINAPI NtContinue( CONTEXT *context, BOOLEAN alertable )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     return NtContinueEx( context, ULongToPtr(alertable) );
 }
 
--- a/wine/dlls/ntdll/unix/sync.c
+++ b/wine/dlls/ntdll/unix/sync.c
@@ -1184,6 +1184,7 @@
 NTSTATUS WINAPI GPT_IMPORT(NtCreateEvent)( HANDLE *handle, ACCESS_MASK access, const OBJECT_ATTRIBUTES *attr,
                                            EVENT_TYPE type, BOOLEAN state )
 {
+    KOPRU_CLEAN_SMALL_ARG( state );
     unsigned int ret;
     data_size_t len;
     struct object_attributes *objattr;
@@ -1434,6 +1435,7 @@
 NTSTATUS WINAPI NtCreateMutant( HANDLE *handle, ACCESS_MASK access, const OBJECT_ATTRIBUTES *attr,
                                 BOOLEAN owned )
 {
+    KOPRU_CLEAN_SMALL_ARG( owned );
     unsigned int ret;
     data_size_t len;
     struct object_attributes *objattr;
@@ -1975,6 +1977,7 @@
 NTSTATUS WINAPI NtWaitForDebugEvent( HANDLE handle, BOOLEAN alertable, LARGE_INTEGER *timeout,
                                      DBGUI_WAIT_STATE_CHANGE *state )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     union debug_event_data data;
     unsigned int ret;
     BOOL wait = TRUE;
@@ -2375,6 +2378,7 @@
 NTSTATUS WINAPI NtSetTimer( HANDLE handle, const LARGE_INTEGER *when, PTIMER_APC_ROUTINE callback,
                             void *arg, BOOLEAN resume, ULONG period, BOOLEAN *state )
 {
+    KOPRU_CLEAN_SMALL_ARG( resume );
     unsigned int ret = STATUS_SUCCESS;
 
     TRACE( "(%p,%p,%p,%p,%08x,0x%08x,%p)\n", handle, when, callback, arg, resume, period, state );
@@ -2472,6 +2476,7 @@
 NTSTATUS WINAPI NtWaitForMultipleObjects( DWORD count, const HANDLE *handles, WAIT_TYPE type,
                                           BOOLEAN alertable, const LARGE_INTEGER *timeout )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     union select_op select_op;
     UINT i, flags = SELECT_INTERRUPTIBLE;
     unsigned int ret;
@@ -2512,6 +2517,7 @@
  */
 NTSTATUS WINAPI NtWaitForSingleObject( HANDLE handle, BOOLEAN alertable, const LARGE_INTEGER *timeout )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     union select_op select_op;
     UINT flags = SELECT_INTERRUPTIBLE;
     unsigned int ret;
@@ -2539,6 +2545,7 @@
 NTSTATUS WINAPI NtSignalAndWaitForSingleObject( HANDLE signal, HANDLE wait,
                                                 BOOLEAN alertable, const LARGE_INTEGER *timeout )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     union select_op select_op;
     UINT flags = SELECT_INTERRUPTIBLE;
     NTSTATUS ret;
@@ -2587,6 +2594,7 @@
  */
 NTSTATUS WINAPI NtDelayExecution( BOOLEAN alertable, const LARGE_INTEGER *timeout )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     unsigned int status = STATUS_SUCCESS;
 
     /* if alertable, we need to query the server */
@@ -2717,6 +2725,7 @@
  */
 NTSTATUS WINAPI NtSetTimerResolution( ULONG res, BOOLEAN set, ULONG *current_res )
 {
+    KOPRU_CLEAN_SMALL_ARG( set );
     static BOOL has_request = FALSE;
 
     TRACE( "(%u,%u,%p), semi-stub!\n", res, set, current_res );
@@ -2844,6 +2853,7 @@
 NTSTATUS WINAPI NtWaitForKeyedEvent( HANDLE handle, const void *key,
                                      BOOLEAN alertable, const LARGE_INTEGER *timeout )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     union select_op select_op;
     UINT flags = SELECT_INTERRUPTIBLE;
 
@@ -2865,6 +2875,7 @@
 NTSTATUS WINAPI NtReleaseKeyedEvent( HANDLE handle, const void *key,
                                      BOOLEAN alertable, const LARGE_INTEGER *timeout )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     union select_op select_op;
     UINT flags = SELECT_INTERRUPTIBLE;
 
@@ -3040,6 +3051,7 @@
 NTSTATUS WINAPI NtRemoveIoCompletionEx( HANDLE handle, FILE_IO_COMPLETION_INFORMATION *info, ULONG count,
                                         ULONG *written, LARGE_INTEGER *timeout, BOOLEAN alertable )
 {
+    KOPRU_CLEAN_SMALL_ARG( alertable );
     HANDLE wait_handle = NULL;
     unsigned int status;
     ULONG i = 0;
@@ -3284,6 +3296,7 @@
 NTSTATUS WINAPI NtAcceptConnectPort( HANDLE *handle, ULONG id, LPC_MESSAGE *msg, BOOLEAN accept,
                                      LPC_SECTION_WRITE *write, LPC_SECTION_READ *read )
 {
+    KOPRU_CLEAN_SMALL_ARG( accept );
     FIXME("(%p,%u,%p,%d,%p,%p),stub!\n", handle, id, msg, accept, write, read );
     return STATUS_NOT_IMPLEMENTED;
 }
@@ -3460,6 +3473,7 @@
  */
 NTSTATUS WINAPI NtDeleteAtom( RTL_ATOM atom )
 {
+    KOPRU_CLEAN_SMALL_ARG( atom );
     unsigned int status;
 
     if (!atom) status = STATUS_INVALID_HANDLE;
@@ -3502,6 +3516,7 @@
 NTSTATUS WINAPI NtQueryInformationAtom( RTL_ATOM atom, ATOM_INFORMATION_CLASS class,
                                         void *ptr, ULONG size, ULONG *retsize )
 {
+    KOPRU_CLEAN_SMALL_ARG( atom );
     unsigned int status;
 
     switch (class)
@@ -3859,6 +3874,7 @@
  */
 NTSTATUS WINAPI NtCommitTransaction( HANDLE transaction, BOOLEAN wait )
 {
+    KOPRU_CLEAN_SMALL_ARG( wait );
     FIXME( "%p, %d stub.\n", transaction, wait );
 
     return STATUS_SUCCESS;
@@ -3869,6 +3885,7 @@
  */
 NTSTATUS WINAPI NtRollbackTransaction( HANDLE transaction, BOOLEAN wait )
 {
+    KOPRU_CLEAN_SMALL_ARG( wait );
     FIXME( "%p, %d stub.\n", transaction, wait );
 
     return STATUS_ACCESS_VIOLATION;
--- a/wine/dlls/ntdll/unix/system.c
+++ b/wine/dlls/ntdll/unix/system.c
@@ -4688,6 +4688,7 @@
 NTSTATUS WINAPI NtInitiatePowerAction( POWER_ACTION action, SYSTEM_POWER_STATE state,
                                        ULONG flags, BOOLEAN async )
 {
+    KOPRU_CLEAN_SMALL_ARG( async );
     FIXME( "(%d,%d,0x%08x,%d),stub\n", action, state, flags, async );
     return STATUS_NOT_IMPLEMENTED;
 }
--- a/wine/dlls/ntdll/unix/thread.c
+++ b/wine/dlls/ntdll/unix/thread.c
@@ -1327,6 +1327,7 @@
                                 HANDLE process, CLIENT_ID *id, CONTEXT *ctx, INITIAL_TEB *teb,
                                 BOOLEAN suspended )
 {
+    KOPRU_CLEAN_SMALL_ARG( suspended );
     FIXME( "%p %d %p %p %p %p %p %d, stub!\n",
            handle, access, attr, process, id, ctx, teb, suspended );
     return STATUS_NOT_IMPLEMENTED;
--- a/wine/dlls/ntdll/unix/unix_private.h
+++ b/wine/dlls/ntdll/unix/unix_private.h
@@ -733,4 +733,12 @@
 
 #endif  /* defined(__i386__) || defined(__x86_64__) */
 
+/* Köprü (patches/crossover-26.3/0005): Windows (MS x64) tarafı 32 bitten küçük bir argümanı
+ * (BOOLEAN, WORD…) geçirirken register'ın/yığın yuvasının yalnızca alt baytlarını yazar; Apple
+ * clang ise argümanın 32 bite genişletilmiş geldiğini varsayıp tamamına bakar. Değeri kendi
+ * genişliğinde volatile belleğe yazıp geri okumak üst baytlardaki çöpü temizler. Syscall olarak
+ * çağrılan fonksiyonların başında kullanılır; liste: tools/abi-scan/scan_small_args.py */
+#define KOPRU_CLEAN_SMALL_ARG(x) \
+    do { volatile __typeof__(x) kopru_small_arg_ = (x); (x) = kopru_small_arg_; } while (0)
+
 #endif /* __NTDLL_UNIX_PRIVATE_H */
--- a/wine/dlls/win32u/class.c
+++ b/wine/dlls/win32u/class.c
@@ -694,6 +694,7 @@
  */
 ULONG WINAPI NtUserGetAtomName( ATOM atom, UNICODE_STRING *name )
 {
+    KOPRU_CLEAN_SMALL_ARG( atom );
     WCHAR buffer[MAX_ATOM_LEN];
     UINT size = 0;
 
@@ -950,6 +951,7 @@
  */
 WORD WINAPI NtUserSetClassWord( HWND hwnd, INT offset, WORD newval )
 {
+    KOPRU_CLEAN_SMALL_ARG( newval );
     return set_class_long_size( hwnd, offset, newval, sizeof(WORD), TRUE );
 }
 
--- a/wine/dlls/win32u/input.c
+++ b/wine/dlls/win32u/input.c
@@ -1095,6 +1095,7 @@
  */
 WORD WINAPI NtUserVkKeyScanEx( WCHAR chr, HKL layout )
 {
+    KOPRU_CLEAN_SMALL_ARG( chr );
     const KBDTABLES *kbd_tables;
     SHORT ret;
 
--- a/wine/dlls/win32u/palette.c
+++ b/wine/dlls/win32u/palette.c
@@ -508,6 +508,7 @@
  */
 HPALETTE WINAPI NtUserSelectPalette( HDC hdc, HPALETTE hpal, WORD bkg )
 {
+    KOPRU_CLEAN_SMALL_ARG( bkg );
     BOOL is_primary = FALSE;
     HPALETTE ret = 0;
     DC *dc;
@@ -626,6 +627,8 @@
 LONG WINAPI NtGdiDoPalette( HGDIOBJ handle, WORD start, WORD count, void *entries,
                             DWORD func, BOOL inbound )
 {
+    KOPRU_CLEAN_SMALL_ARG( start );
+    KOPRU_CLEAN_SMALL_ARG( count );
     switch (func)
     {
     case NtGdiAnimatePalette:
--- a/wine/dlls/win32u/win32u_private.h
+++ b/wine/dlls/win32u/win32u_private.h
@@ -429,4 +429,12 @@
     return !IsRectEmpty( dst );
 }
 
+/* Köprü (patches/crossover-26.3/0005): Windows (MS x64) tarafı 32 bitten küçük bir argümanı
+ * (BOOLEAN, WORD…) geçirirken register'ın/yığın yuvasının yalnızca alt baytlarını yazar; Apple
+ * clang ise argümanın 32 bite genişletilmiş geldiğini varsayıp tamamına bakar. Değeri kendi
+ * genişliğinde volatile belleğe yazıp geri okumak üst baytlardaki çöpü temizler. Syscall olarak
+ * çağrılan fonksiyonların başında kullanılır; liste: tools/abi-scan/scan_small_args.py */
+#define KOPRU_CLEAN_SMALL_ARG(x) \
+    do { volatile __typeof__(x) kopru_small_arg_ = (x); (x) = kopru_small_arg_; } while (0)
+
 #endif /* __WINE_WIN32U_PRIVATE */
--- a/wine/dlls/win32u/window.c
+++ b/wine/dlls/win32u/window.c
@@ -1580,6 +1580,7 @@
  */
 BOOL WINAPI NtUserSetWindowFNID( HWND hwnd, WORD fnid )
 {
+    KOPRU_CLEAN_SMALL_ARG( fnid );
     int off = FNID_OFF(fnid);
     int len = FNID_LEN(fnid);
     WND *win;
@@ -1629,6 +1630,7 @@
  */
 WORD WINAPI NtUserSetWindowWord( HWND hwnd, INT offset, WORD newval )
 {
+    KOPRU_CLEAN_SMALL_ARG( newval );
     if (offset < 0 && offset != GWLP_USERDATA)
     {
         RtlSetLastWin32Error( ERROR_INVALID_INDEX );
@@ -2607,6 +2609,7 @@
  */
 BOOL WINAPI NtUserSetLayeredWindowAttributes( HWND hwnd, COLORREF key, BYTE alpha, DWORD flags )
 {
+    KOPRU_CLEAN_SMALL_ARG( alpha );
     BOOL ret;
 
     TRACE( "(%p,%s,%d,%x)\n", hwnd, debugstr_color(key), alpha, flags );
